### PR TITLE
Fix SDR amount in prepare output

### LINKED ISSUE #1905: [Bug]: When sending dust transactions which results in a storage deposit <42600 glow
## Bug description

Currently, it's not possible to prepare an output that sends an amount between 8001-42600 for basic transactions (this range changes once transactions features are added). When sending these outputs the storage deposit return transaction doesn't cover the minimum required storage deposit and thus rejects the transaction.

## Version

Which version of the library are you using?

- rc-22

## Steps To reproduce the bug

Explain how the maintainer can reproduce the bug.

Run the following example: 

```
/**
 * This example builds an output.
 */

const getUnlockedManager = require('./account-manager');

async function run() {
    try {
        const manager = await getUnlockedManager();
        const account = await manager.getAccount('Alice');

        //TODO: Replace with the address of your choice!
        const recipientAddress =
            'rms1qrrv7flg6lz5cssvzv2lsdt8c673khad060l4quev6q09tkm9mgtupgf0h0';
        const amount = '8000';
        const output = await account.prepareOutput({
            recipientAddress,
            amount,
        });
        const minimumRequiredStorageDeposit =
            await account.minimumRequiredStorageDeposit(output);

        console.log('Output:', output);
        console.log(
            'Minimum required storage deposit:',
            minimumRequiredStorageDeposit,
        );
        const transaction = await account.sendOutputs([output]);
        console.log('TX:', transaction)
    } catch (error) {
        console.log('Error: ', error);
    }
    process.exit(0);
}

run();
```

## Expected behaviour

The storage deposit should be increased so that it covers the return unlock condition at all times.

## Actual behaviour

Wallet.rs errors

## Errors

`the return deposit (8001) must be greater than the minimum storage deposit (42600)`

### PR DESCRIPTION
**Pull request by Thoralf-M**
# Description of change

Fix SDR amount in prepare output

## Links to any relevant issues

Fixes https://github.com/iotaledger/wallet.rs/issues/1905

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Added test

## Change checklist

- [ ] I have followed the contribution guidelines for this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that new and existing unit tests pass locally with my changes

### CODE REVIEW

File: wallet/tests/output_preparation.rs [hunk 1]
```diff
@@ -370,6 +370,52 @@ async fn output_preparation() -> Result<()> {
...
+    let rent_structure = account.client().get_rent_structure().await?;
+    let token_supply = account.client().get_token_supply().await?;
+    // Check if the output has enough amount to cover the storage deposit
+    output.verify_storage_deposit(rent_structure, token_supply)?;
+    assert_eq!(output.amount(), 85199);
+    // address and sdr unlock condition
+    assert_eq!(output.unlock_conditions().unwrap().len(), 2);
+    let sdr = output.unlock_conditions().unwrap().storage_deposit_return().unwrap();
+    assert_eq!(sdr.amount(), 42600);
+
```
Comment (thibault-martinez) [inline wallet/tests/output_preparation.rs:429]: If you're testing a specific scenario, I'd isolate it in a dedicated test, easier to read

File: wallet/tests/output_preparation.rs [hunk 2]
```diff
@@ -370,6 +370,52 @@ async fn output_preparation() -> Result<()> {
...
 
+    let output = account
+        .prepare_output(
+            OutputOptions {
+                recipient_address: recipient_address.clone(),
+                amount: 8001,
+                assets: None,
+                features: None,
+                unlocks: None,
+                storage_deposit: None,
```
Comment (thibault-martinez) [inline wallet/tests/output_preparation.rs:396]: What if there is already an sdr? Maybe add a test?

File: wallet/src/account/operations/transaction/prepare_output.rs [hunk 3]
```diff
@@ -116,12 +119,7 @@ impl AccountHandle {
 
                     // Calculate the minimum storage deposit to be returned
                     let min_storage_deposit_return_amount =
-                        BasicOutputBuilder::new_with_minimum_storage_deposit(rent_structure.clone())?
-                            .add_unlock_condition(UnlockCondition::Address(AddressUnlockCondition::new(
-                                Address::try_from_bech32(options.recipient_address.clone())?.1,
-                            )))
-                            .finish_output(token_supply)?
-                            .amount();
+                        minimum_storage_deposit_basic_output(&rent_structure, &None, token_supply)?;
```
Comment (thibault-martinez) [inline wallet/src/account/operations/transaction/prepare_output.rs:122]: Is that just for convenience or did it not return the same value?

File: wallet/src/account/operations/transaction/prepare_output.rs [hunk 4]
```diff
@@ -167,9 +160,19 @@ impl AccountHandle {
         // We might have added more unlock conditions, so we check the minimum storage deposit again and update the
         // amounts if needed
         if second_output.amount() < required_storage_deposit {
-            third_output_builder = third_output_builder.with_amount(required_storage_deposit)?;
+            let mut new_sdr_amount = required_storage_deposit - options.amount;
```
Comment (thibault-martinez) [inline wallet/src/account/operations/transaction/prepare_output.rs:163]: Bit annoying to check in the diff but is `second_output.amount()` == `options.amount` ?

File: wallet/src/account/operations/transaction/prepare_output.rs (hunk 3 above)
Comment (Thoralf-M) [reply wallet/src/account/operations/transaction/prepare_output.rs:122]: Just convenience

File: wallet/src/account/operations/transaction/prepare_output.rs (hunk 4 above)
Comment (Thoralf-M) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: It could be, but doesn't have to be

Comment (thibault-martinez) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: So nothing guarantees that `required_storage_deposit` > `options.amount` ? This could underflow ?

File: wallet/src/account/operations/transaction/prepare_output.rs [hunk 5]
```diff
@@ -167,9 +160,19 @@ impl AccountHandle {
...
+            let minimum_storage_deposit = minimum_storage_deposit_basic_output(&rent_structure, &None, token_supply)?;
+            let mut final_output_amount = required_storage_deposit;
+            if required_storage_deposit - options.amount < minimum_storage_deposit {
```
Comment (thibault-martinez) [inline wallet/src/account/operations/transaction/prepare_output.rs]: Maybe do the opposite check to avoid the underflow ?

File: wallet/tests/output_preparation.rs (hunk 2 above)
Comment (Thoralf-M) [reply wallet/tests/output_preparation.rs:396]: You can't really provide it, it's just an option to set if you want a sdr unlock condition (default) or if you want to gift the storage deposit
But added a test for that

File: wallet/src/account/operations/transaction/prepare_output.rs (hunk 4 above)
Comment (Thoralf-M) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: `second_output.amount()` >= `options.amount`, so `second_output.amount() < required_storage_deposit` in 162 should take care of that

File: wallet/src/account/operations/transaction/prepare_output.rs (hunk 5 above)
Comment (Thoralf-M) [reply wallet/src/account/operations/transaction/prepare_output.rs]: Changed to prevent errors in the future if we want to change something, but it's safe because of `if second_output.amount() < required_storage_deposit {`

File: wallet/CHANGELOG.md [hunk 6]
```diff
@@ -39,6 +39,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   contains a `ParticipationEventRegistrationOptions` and expects a `ParticipationEvents` response;
 - `AccountMethod::GetParticipationEventIds` now also contains a `Node`;
 - Custom `Serialize` implementation for `Error`;
+- Made `TransactionOptionsDto, ReturnStrategy, MintTokenTransactionDto` pub reachable;
```
Comment (thibault-martinez) [inline wallet/CHANGELOG.md]:
```suggestion
- Made `TransactionOptionsDto`, `ReturnStrategy` and `MintTokenTransactionDto` pub reachable;
```

File: wallet/src/account/operations/transaction/prepare_output.rs (hunk 4 above)
Comment (thibault-martinez) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: Where is the `second_output.amount() >= options.amount`  guarantee?

Comment (thibault-martinez) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: Oj because it's used to create the builder I assume

Comment (thibault-martinez) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: It's very hard to follow if it's guaranteed, I'll trust you

Comment (Thoralf-M) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: https://github.com/iotaledger/wallet.rs/blob/f5578dc0b71ab9e04c1cec1c3e7ec5197c7b79fa/wallet/src/account/operations/transaction/prepare_output.rs#L109-L111

Comment (Thoralf-M) [reply wallet/src/account/operations/transaction/prepare_output.rs:163]: Ordering::Less => https://github.com/iotaledger/wallet.rs/blob/f5578dc0b71ab9e04c1cec1c3e7ec5197c7b79fa/wallet/src/account/operations/transaction/prepare_output.rs#L147

**Review by thibault-martinez — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]
